WalletWasabi/Services/SmartCoinSelector.cs

```diff
@@ -0,0 +1,89 @@
+using NBitcoin;
+using System.Linq;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using WalletWasabi.Models;
+using WalletWasabi.Exceptions;
+
+namespace WalletWasabi.Services
+{
+	public class SmartCoinSelector : ICoinSelector
+	{
+		private Dictionary<OutPoint, SmartCoin> _smartsCoinByOutpoint;
+
+		public SmartCoinSelector(Dictionary<OutPoint, SmartCoin> smartCoinsByOutpoint)
+		{
+			_smartsCoinByOutpoint = smartCoinsByOutpoint;
+		}
+
+		public IEnumerable<ICoin> Select(IEnumerable<ICoin> coins, IMoney target)
+		{
+			var coinsByOutpoint = coins.ToDictionary(c => c.Outpoint);
+			var totalOutAmount = (Money)target;
+			var unspentCoins = coins.Select(c => _smartsCoinByOutpoint[c.Outpoint]).ToArray();
+			var coinsToSpend = new HashSet<SmartCoin>();
+			var unspentConfirmedCoins = new List<SmartCoin>();
+			var unspentUnconfirmedCoins = new List<SmartCoin>();
+			foreach (SmartCoin coin in unspentCoins)
+			{
+				if (coin.Confirmed)
+				{
+					unspentConfirmedCoins.Add(coin);
+				}
+				else
+				{
+					unspentUnconfirmedCoins.Add(coin);
+				}
+			}
+
+			bool haveEnough = TrySelectCoins(ref coinsToSpend, totalOutAmount, unspentConfirmedCoins);
+			if (!haveEnough)
+			{
+				haveEnough = TrySelectCoins(ref coinsToSpend, totalOutAmount, unspentUnconfirmedCoins);
+			}
+
+			if (!haveEnough)
+			{
+				throw new InsufficientBalanceException(totalOutAmount, unspentConfirmedCoins.Select(x => x.Amount).Sum() + unspentUnconfirmedCoins.Select(x => x.Amount).Sum());
```

Comment:
probably, I copy/pasta the previous code to facilitate review.

Reply:
```
unspentConfirmedCoins.Concat(unspentUnconfirmedCoins).Sum(x => x.Amount)
```
This is incorrect, because this will returns a `decimal` while my current version returns a `Money`